src/index.ts

```diff
@@ -16,7 +16,7 @@ export type ValidatedType<
         : O extends Array<'boolean'>
         ? boolean[] | undefined
         : O extends ObjectShape<O>
-        ? { [key in keyof O]: ValidatedType<O[key]> } | undefined
+        ? MapObjectShape<O> | undefined
         : O extends Array<infer U>
         ? U extends ObjectShape<T>
             ? Array<{ [key in keyof U]: ValidatedType<U[key]> }> | undefined
```

Comment:
Would switching to using MapObjectShape<T> here work?
```suggestion
            ? Array<MapObjectShape<U>> | undefined
```

Reply:
Fixed